pkg/controller/externalnode/controller_test.go

```diff
@@ -0,0 +1,606 @@
+				return checkExternalEntityExists(controller.crdClient, tc.expectedEntity)
+			})
+			assert.NoError(t, err)
+			if tc.existingEntity != nil {
+				exists, err := checkExternalEntityExists(controller.crdClient, tc.existingEntity)
+				assert.NoError(t, err)
+				assert.False(t, exists)
+			}
+		})
+
+	}
+}
+
+func TestDeleteExternalNode(t *testing.T) {
+	externalNode := &v1alpha1.ExternalNode{
+		ObjectMeta: metav1.ObjectMeta{Name: "vm1", Namespace: "ns1", Labels: map[string]string{"en": "vm1"}},
+		Spec: v1alpha1.ExternalNodeSpec{
+			Interfaces: []v1alpha1.NetworkInterface{{IPs: []string{"1.1.1.2"}}},
+		},
+	}
+	expectedEntity := &v1alpha2.ExternalEntity{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      "vm1",
+			Namespace: "ns1",
+			OwnerReferences: []metav1.OwnerReference{
+				{
+					APIVersion: "crd.antrea.io/v1alpha1",
+					Kind:       "ExternalNode",
+					Name:       "vm1",
+				},
+			},
+			Labels: map[string]string{"en": "vm1"},
+		},
+		Spec: v1alpha2.ExternalEntitySpec{
+			Endpoints: []v1alpha2.Endpoint{
+				{IP: "1.1.1.2"},
+			},
+			ExternalNode: "vm1",
+		},
+	}
+	testDeleteExternalNode(t, externalNode, expectedEntity)
+}
+
+func TestReconcileExternalNodes(t *testing.T) {
+	existingExternalNode := &v1alpha1.ExternalNode{
+		ObjectMeta: metav1.ObjectMeta{Name: "vm3", Namespace: "ns1", Labels: map[string]string{"en": "vm3"}},
+		Spec: v1alpha1.ExternalNodeSpec{
+			Interfaces: []v1alpha1.NetworkInterface{{IPs: []string{"1.1.1.4", "2.2.2.4"}}},
+		},
+	}
+	existingEntity := &v1alpha2.ExternalEntity{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      "vm1",
+			Namespace: "ns1",
+			OwnerReferences: []metav1.OwnerReference{
+				{
+					APIVersion: "crd.antrea.io/v1alpha1",
+					Kind:       "ExternalNode",
+					Name:       "vm1",
+				},
+			},
+			Labels: map[string]string{"en": "vm1"},
+		},
+		Spec: v1alpha2.ExternalEntitySpec{
+			Endpoints: []v1alpha2.Endpoint{
+				{IP: "1.1.1.4"},
+			},
+			ExternalNode: "vm1",
+		},
+	}
+	expectedEntity := &v1alpha2.ExternalEntity{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      "vm3",
+			Namespace: "ns1",
+			OwnerReferences: []metav1.OwnerReference{
+				{
+					APIVersion: "crd.antrea.io/v1alpha1",
+					Kind:       "ExternalNode",
+					Name:       "vm3",
+				},
+			},
+			Labels: map[string]string{"en": "vm3"},
+		},
+		Spec: v1alpha2.ExternalEntitySpec{
+			Endpoints: []v1alpha2.Endpoint{
+				{IP: "1.1.1.4"},
+				{IP: "2.2.2.4"},
+			},
+			ExternalNode: "vm3",
+		},
+	}
+	crdObjects := []runtime.Object{existingExternalNode, existingEntity}
+	controller := newExternalNodeController(crdObjects)
+	stopCh := make(chan struct{})
+	defer close(stopCh)
+	informerFactory.Start(stopCh)
+	informerFactory.WaitForCacheSync(stopCh)
+	err := controller.reconcileExternalNodes()
+	require.NoError(t, err)
+	err = wait.PollImmediate(time.Millisecond*50, time.Second, func() (done bool, err error) {
```

Comment:
Unlike other tests that rely on `Run` method to create/delete objects asynchronously, `reconcileExternalNodes` does everything synchronously, I think it should do the following check immediately and one-time?

Reply:
updated.